genegram/model.py

```diff
@@ -0,0 +1,280 @@
+            + str(pair)
+            + "    "
+            + str(i + 1)
+            + "\n"
+        )
+    return ct
+
+
+def remove_multiplets(img):
+    def get_multiplets(i0, j0, img):
+        mps = []
+        size = len(img)
+        for i in range(size):
+            if img[i0, i] == 255 and (i0, i) != (i0, j0) and i0 <= i:
+                mps.append((i0, i))
+            if img[j0, i] == 255 and (j0, i) != (i0, j0) and j0 <= i:
+                mps.append((j0, i))
+            if img[i, i0] == 255 and (i, i0) != (i0, j0) and i <= i0:
+                mps.append((i, i0))
+            if img[i, j0] == 255 and (i, j0) != (i0, j0) and i <= j0:
+                mps.append((i, j0))
+        return list(set(mps))
+
+    def get_stem_len(i0, j0, img):
+        size = len(img)
+        stem_len = 1
+        i, j = i0 + 1, j0 - 1
+        while (
+            i < len(img)
+            and j >= 0
+            and img[i][j] == 255
+            and len(get_multiplets(i, j, img)) == 0
+        ):
+            stem_len += 1
+            i += 1
+            j -= 1
+        i, j = i0 - 1, j0 + 1
+        while (
+            i >= 0
+            and j < len(img)
+            and img[i][j] == 255
+            and len(get_multiplets(i, j, img)) == 0
+        ):
+            stem_len += 1
+            i -= 1
+            j += 1
+        return stem_len
+
+    size = len(img)
+    mps_nums = dict()
+    for i in range(size):
+        for j in range(i + 1, size):
+            if img[i][j] == 255:
+                mps = get_multiplets(i, j, img)
+                if len(mps) > 0:
+                    mps_nums[(i, j)] = len(mps)
+    mps_nums = {k: v for k, v in sorted(mps_nums.items(), key=lambda item: -item[1])}
+    to_delete = []
+    while len(mps_nums) > 0:
+        for el in to_delete:
+            del mps_nums[el]
+        to_delete = []
+        for (i, j) in mps_nums.keys():
+            if not (i, j) in to_delete:
+                mps = get_multiplets(i, j, img)
+                if len(mps) > 0:
+                    min_stem_len = get_stem_len(i, j, img)
+                    i_pick, j_pick = i, j
+                    for (i0, j0) in mps:
+                        stem_len = get_stem_len(i0, j0, img)
+                        if stem_len < min_stem_len:
+                            i_pick, j_pick = i0, j0
+                            min_stem_len = stem_len
+                    img[i_pick][j_pick] = 0
+                    to_delete.append((i_pick, j_pick))
+                else:
+                    to_delete.append((i, j))
+    return img
+
+
+# Model definition functions
+
+# layer that for inputs i1, i2, i3, i4
+# returns (w1*i1 + w2*i2 + w3*i3 + w4*i4) / 4, where wi are trainable coeffs
+class WeightedSum(Layer):
+    def __init__(self):
+        super(WeightedSum, self).__init__()
+        w_init = tf.keras.initializers.Ones()
+        self.w1 = tf.Variable(
+            initial_value=w_init(shape=(), dtype="float32"), trainable=True
```

Comment:
Why layers are trainable? I guess, model training and model using should be separated.

Reply:
Unfortunately, I don't know how to do it better yet. @LuninaPolina, can you help? :eyes: